### DISCUSSION TITLE
Results equality across different platforms

### DISCUSSION THREAD

**Discussion by olologin** [shibatch/sleef#482] [Q&A] [answered]
Hi everyone,
I work on a very specific project that requires trigonometric results to be equal among different platforms (x64 Windows/Linux/Mac and aarch64 Linux/Mac are supported now).
I was investigating available libraries that can be helpful for this, and I was investigating different fdlibm successors.
I ended up adapting openlibm for our codebase, but today my coworker told me about your project, and it also seems good for us.

**Question**: Is it possible to build sleef for all of our platforms so that it gives equal results for trigonometric functions everywhere?
I understand that a lot depends on a compiler and its optimizer, but lets ignore compiler-related risks and just talk about sleef implementation. I see you are using a lot of vector instructions, do these produce equal results on ARM and x64? In our case we can sacrifice performance for reproducibility of results, so maybe we can somehow configure sleef to be a bit slower but give equal results everywhere?

**Comment by blapie** [comment, marked as answer]
Hello! Thanks for starting a discussion on this, this is an important topic in SLEEF.

This topic has been discussed in several issues so far under various terminology, such as determinism, bitwise reproducibility, consistency,... As you might know there are various declinations of it, reproducibility can be required across architectures/platforms or simply between scalar and vector implementations (or even across vector extensions).
I encourage you to browse through existing issues (https://github.com/shibatch/sleef/issues?q=is%3Aopen+is%3Aissue+label%3Adeterminism), to see if one is more appropriate to your use case.

For reproducibility across platforms and vector extensions (ISAs), availability of FMA instructions is what matters most. If some of your platforms don't have FMA enabled/available, then you will need to disable FMAs on all ISAs to ensure reproducibility, which will have a significant impact on performance (but you said you might be ok with it). Similarly, FTZ is not a feature that is consistent across platform (with impact on float to integer conversions).

Determinism was added as a feature by this PR https://github.com/shibatch/sleef/pull/204, and the corresponding implementations have finz or cinz in the name. Looks like you are lucky because most important routines (including trigonometric routines) support it, and they exist in different precisions and accuracies too. Here is an exhaustive list: https://github.com/shibatch/sleef/blob/69cfa7130f3f2bda838633071b1c63aa7d56f940/src/libm-tester/hash_finz.txt#L4   

If you need scalar to vector determinism then these routines are also available in scalar flavour (as discussed here https://github.com/shibatch/sleef/issues/443 or https://github.com/shibatch/sleef/issues/444).

This is certainly something we want to document more accurately in the future. Let us know if you are facing difficulties using them and please report a bug if you notice some flaws in the implementations. 

Regards,

**Reply by olologin** [reply]
Wow. Thanks for quick answer. You even have tests specifically for this request.
I will test sleef :)

**Comment by blapie** [comment]
@xoofx I know you have been involved in introducing this feature, feel free to complete my answer if I missed something. Cheers!
